src/type_inf.jl

```diff
@@ -56,6 +56,343 @@ function infer_type(binding::Binding, scope, state)
+        return get_struct_fieldname(x[1])
+    elseif typof(x) === CSTParser.InvisBrackets && length(x) == 3
+        return get_struct_fieldname(x[2])
+    elseif isidentifier(x)
+        return x
+    else
+    end
+    return nothing
+end
+
+function cst_struct_fieldnames(x::EXPR)
+    fns = Symbol[]
+    if CSTParser.defines_mutable(x)
+        body = x[4]
+    elseif CSTParser.defines_struct(x)
+        body = x[3]
+    else
+        return fns
+    end
+    for arg in body
+        field_name = get_struct_fieldname(arg)
+        if field_name isa EXPR && isidentifier(field_name)
+            push!(fns, Symbol(CSTParser.str_value(field_name)))
+        end
+    end
+    return fns
+end
+
+
+"""
+    fieldname_type_map(s::Union{Scope,ModuleStore,EnvStore}, server, l = Dict())
+
+Returns a Dict where a fieldname (key) points to a collection of types that 
+have that field. 
+"""
+fieldname_type_map(s, server, l = Dict{Symbol,Any}()) = l # fallback
+function fieldname_type_map(s::Scope, server, l = Dict())
+    for (n,b) in s.names
+        b = get_root_method(b, server)
+        # Todo: Allow for const rebindings of datatypes (i.e. `const dt = DataType`)
+        if b isa Binding && b.val isa EXPR 
+            if CSTParser.defines_datatype(b.val)
+                for f in cst_struct_fieldnames(b.val)
+                    f = Symbol(f)
+                    if haskey(l, f)
+                        push!(l[f], b)
+                    else
+                        l[f] = [b]
+                    end
+                end
+            elseif CSTParser.defines_function(b.val) && n == "get_property"
+                # need to check this overwrites Base.get_property
+                # need to iterate over all methods
+                sig = CSTParser.get_sig(b.val)
+                if length(sig) > 5 && _binary_assert(sig[3], CSTParser.Tokens.DECLARATION) && hasref(sig[3][3])
+                    t_binding = refof(sig[3][3])
+                    if t_binding isa Binding
+                        if t_binding.type !== CoreTypes.DataType
+                            t_binding = get_root_method(t_binding, server)
+                            t_binding.type !== CoreTypes.DataType && continue
+                        end
+                        for f in get_property_shadow_fields(b.val)
+                            f = Symbol(f)
+                            if haskey(l, f)
+                                push!(l[f], t_binding)
+                            else
+                                l[f] = [t_binding]
+                            end
+                        end
+                    end
+                end
+            end
+        end
+    end
+    return l
+end
+
+"""
+    get_property_shadow_fields(func)
+
+Assumes `func` is the definition of a function for `get_property`. Searches for 
+comparisons within the body between the second argument of the function and
+symbols, returning a list of these symbols.
```

Comment:
Is that a common pattern? The cases of `getproperty` that I'm familiar with almost all look in a dict, or some type parameters or something like that for names...